keps/sig-cli/3104-introduce-kuberc/README.md

```diff
@@ -0,0 +1,926 @@
+
+1. [How are subcommands indicated for command overrides?](https://github.com/kubernetes/enhancements/pull/3392#discussion_r896174406)
+1. [How are subcommand aliases indicated?](https://github.com/kubernetes/enhancements/pull/3392#discussion_r896179267)
+1. [How do we handle tying these settings to cluster contexts?](https://github.com/kubernetes/enhancements/pull/3392#discussion_r898239057)
+1. [Do we want this file to live elsewhere i.e. XDG_CONFIG?](https://github.com/kubernetes/enhancements/pull/3392#discussion_r896177353)
+1. [How do we exectue subcommands and do we want to support variable substitution i.e. `$1`](https://github.com/kubernetes/enhancements/pull/3392#discussion_r898227148)
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+| Risk | Impact | Mitigation |
+| --- | --- | --- |
+| Confusing users with a new config file | Low | Documentation and education |
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+During alpha this feature will be enabled through the `ENABLE_KUBERC=true` environment variable. The file will default to being located in `~/.kube/kuberc`. A flag will allow overriding this default location with a path i.e. `kubectl --kuberc /var/kube/rc`.
+
+Three initial top level keys are proposed.
+
+* `apiVersion` to determine the version of the config.
+* `kind` to keep consistency with Kubernetes resources.
+* `command`
+  * `aliases` for users to declare their own aliases for commands with flags and values.
+  * `overrides` for users to set default flags to apply to commands.
+
+`command.aliases` will not be allowed to override builtins but take precedence of plugins i.e. builtins -> aliases -> plugins. Additional flags and values will be appended to the end of the aliased command. It is the responsibility of the user to define aliases with this in mind.
+
+`command.overrides` is modeled after all configurable behavior being implemented as flags first. This is a design decision that was made after modeling out the intended behavior and realizing that targeting flags filled the use cases. A merge will be done in the execution of the command for flags with inline overrides taking precedence.
+
+```
+apiVersion: v1alpha1
+kind: Preferences
+
+command:
+  aliases:
+    - alias: getdbprod
+      command: get pods -l what=database --namespace us-2-production
+      
+  overrides:
+    - command: apply
+      flags:
```

Comment:
I wouldn't call these flags. https://superuser.com/a/1070071 (and the common use, at least when I was learning computers) was to reserve the term _flag_ for boolean and maybe tristate options. 

I know Golang coders often use the term to mean any kind of command line option, but we shouldn't assume that every tool and user consuming a kuberc is written in Go or knows this Go quirk.